src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedNewTests.cs

```diff
@@ -1124,5 +1331,2092 @@ public static void Main()
+        {
+            string source = @"
+using System;
+
+class C
+{
+    public static bool operator ==(C o1, C o2)
+    {
+        Console.WriteLine(""operator =="");
+        return (object)o1 == (object)o2;
+    }
+    public static bool operator !=(C o1, C o2)
+    {
+        Console.WriteLine(""operator !="");
+        return (object)o1 != (object)o2;
+    }
+
+    static void Main()
+    {
+        Console.WriteLine(new C() == new());
+        Console.WriteLine(new() == new C());
+        Console.WriteLine(new C() != new());
+        Console.WriteLine(new() != new C());
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe.WithWarningLevel(0));
+            comp.VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: @"
+operator ==
+False
+operator ==
+False
+operator !=
+True
+operator !=
+True");
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+            var nodes = tree.GetCompilationUnitRoot().DescendantNodes();
+
+            validate(1, "new()", type: "C", convertedType: "C", symbol: "C..ctor()", 
+                operatorSymbol: "System.Boolean C.op_Equality(C o1, C o2)", ConversionKind.Identity);
+
+            validate(2, "new()", type: "C", convertedType: "C", symbol: "C..ctor()", 
+                operatorSymbol: "System.Boolean C.op_Equality(C o1, C o2)", ConversionKind.Identity);
+
+            validate(5, "new()", type: "C", convertedType: "C", symbol: "C..ctor()", 
+                operatorSymbol: "System.Boolean C.op_Inequality(C o1, C o2)", ConversionKind.Identity);
+
+            validate(6, "new()", type: "C", convertedType: "C", symbol: "C..ctor()", 
+                operatorSymbol: "System.Boolean C.op_Inequality(C o1, C o2)", ConversionKind.Identity);
+
+            void validate(int index, string expression, string type, string convertedType, string symbol, string operatorSymbol, ConversionKind conversionKind)
+            {
+                var @new = nodes.OfType<ObjectCreationExpressionSyntax>().ElementAt(index);
+                Assert.Equal(expression, @new.ToString());
+                Assert.Equal(type, model.GetTypeInfo(@new).Type.ToTestDisplayString());
+                Assert.Equal(convertedType, model.GetTypeInfo(@new).ConvertedType.ToTestDisplayString());
+                Assert.Equal(symbol, model.GetSymbolInfo(@new).Symbol.ToTestDisplayString());
+                Assert.Equal(model.GetConversion(@new).Kind, conversionKind);
+                Assert.Equal(operatorSymbol, model.GetSymbolInfo(@new.Parent).Symbol.ToTestDisplayString());
+            }
+        }
+
+        [Fact]
+        public void TestEquality_Class_UserDefinedOperator_ErrorCases()
+        {
+            string source = @"
+using System;
+
+class D
+{
+}
+
+class C
+{
+    public extern static bool operator ==(C o1, C o2);
+    public extern static bool operator !=(C o1, C o2);
+    public exte
```

Comment:
I'm done with the review pass. @gafter @dotnet/roslyn-compiler for a second review.

This PR resolves many open issues and adds a lot of test coverage. It's awesome. Thanks!

@alrz There are a few unaddressed comments left (mostly test suggestions). Let me know I should ping on those that are older to bring them back to the top. Sorry it took a while to push this code review forward.

Thinking more broadly about completing the feature, are there known holes besides tuple equality and the issue with natural type (should be null) I mentioned above?

I will start refreshing the test plan (https://github.com/dotnet/roslyn/issues/28489), reviewing it with the team (to brainstorm more test ideas), and checking off what is already validated. I'll also do an IDE test pass.
In terms of test coverage, it would be good to add two IOperations tests (one with failed target-typing, and one positive one).
We're aiming to merge the feature back to a trunk branch before January, so that it is available in dev16 preview2 :-)

Reply:
>are there known holes besides tuple equality and the issue with natural type (should be null) I mentioned above? 

Only the syntax public API. See the latest comment in #28489. I think we'd need to finalize that before preview.